### ISSUE TITLE
incorrect update Tree

### ISSUE THREAD

**Issue by xFloooo** [typeorm/typeorm#2418]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

tree-entities ("materialized-path") - incorrect update query
```
[Node] query: START TRANSACTION
[Node] query: UPDATE `auth`.`role` SET `parentId` = ? WHERE `id` = ? -- PARAMETERS: [1,3]
[Node] query: COMMIT
```
- query not updatet field `mpath`

1 - Create Entity
```ts
import {
    Entity,
    Tree,
    Column,
    PrimaryGeneratedColumn,
    TreeChildren,
    TreeParent,
    TreeLevelColumn,
    ManyToMany,
    JoinTable
} from "typeorm";

@Entity({ database: "auth" })
@Tree("materialized-path")
export class Role {
    @PrimaryGeneratedColumn() id: number;

    @Column()
    name: string;

    @Column({
        nullable: true
    })
    description: string;

    @TreeChildren() children: Role[];

    @TreeParent() parent: Role;
}

```

2 - Create 3 Record
```ts
        const treeRepositoryRole = getTreeRepository(Role, "auth");
        const role1 = new Role(); // id 1
        role1.name = 'admin'
        const role2 = new Role(); // id 2
        role2.name = 'Subadmin'
        const role3 = new Role(); // id 3
        role3.name = 'SubSubadmin'
        await repositoryRole.save(role1);
        await repositoryRole.save(role2);
        await repositoryRole.save(role3);
```

3 - update 1 record
```ts
   const treeRepositoryRole = getTreeRepository(Role, "auth");
   const role = await repositoryRole.findOne(3);
   const roleParent = await repositoryRole.findOne(1);
   role.parent = roleParent;
   await treeRepositoryRole.save(role);
``` 
4 - console logs
```
[Node] query: START TRANSACTION
[Node] query: UPDATE `auth`.`role` SET `parentId` = ? WHERE `id` = ? -- PARAMETERS: [1,3]
[Node] query: COMMIT
```
query not updated `mpath` field

**Comment by DaLiangLee** [comment]
How do you solve it? thanks.

**Comment by Kampfmoehre** [comment]
For future travellers I have come up with my own code to update the mpath field.
```typescript
    await this.manager.transaction(async (man: EntityManager) => {
      // if the entity parent does not match the current parent, move it
      // TODO: wait for typeorm to implement tree moving. In the meantime implement our own mechanism
      // see https://github.com/typeorm/typeorm/issues/4174
      if (userfile.parent?.id !== parent.id) {
        // We need to update the mpath of the current object and all children. Something like this query
        // UPDATE file_system_object SET mpath = REPLACE(mpath, 'currentPath', 'newPath') WHERE mpath like '%currentId.%';

        // since typeorm won't give us the mpath column we have to select it ourself :-(
        const pathResult: { mpath: string } = await man
          .createQueryBuilder(FileSystemObject, "fso")
          .select("mpath")
          .where("fso.id = :id", { id: userfile.id })
          .getRawOne();

        // now replace the old parent id with the new parent id and update everyone that has our ID in its path
        const newPath = pathResult.mpath.replace(userfile.parent!.id, parent.id);
        const result = (await man.query(
          `UPDATE file_system_object SET mpath = REPLACE(mpath, '${pathResult.mpath}', '${newPath}') WHERE mpath like '%${userfile.id}.%'`,
        )) as number[];

        // if not at least one entry (the very object we're moving) was updated, then the query or the data is wrong
        if (result[1] < 1) {
          throw new Error("Something went wrong during file move.");
        }

        // now update the parentId field
        userfile.parent = parent;
      }

      await man.getTreeRepository(FileSystemObject).save(userfile);
    });
```
I left the comments to indicate why something is happen. Note that we use UUID as id for the entity but it should be the same with regular numbers.

As stated already in https://github.com/typeorm/typeorm/issues/4232 there is no way to make typeorm return the mpath with the entity, hence the (somewhat ugly) own query to retrieve it.

Also be careful and check if your database supports the REPLACE function (we use postgres but I saw it working in sqlite also).

I could not make the update query work with the UpdateQueryBuilder since it was throwing some errors for me. I tried something like this:
```typescript
await man.createQueryBuilder()
  .update(FileSystemObject)
  .set({ mpath: () => `REPLACE(mpath, '${pathResult.mpath}', '${newPath}')`})
  .where("...")
  .execute();
```

So I had to write a direct update query. Be careful as you have to do your own escaping of the values here to prevent SQL injection!

Of course this is all very ugly. We have to write a plain query though we already loaded the entity only to retrieve a single field. Also update is very difficult, since the mpath field can not be in the entity we have to update it ourselves again via direct query and last we also need to update the parent ID. At least this can do typeorm for us, but it is another query and it's all not very optimised. Consider this more the quick and dirty version and adapt to your needs.

Also don't hesitate to share your improvements here for other to adapt.

TL:DR
Find all rows that have the ID of the moving item in their mpath and replace the ID of the old parent with the new parent. This should cover the complete subtree of the moving item.

**Comment by davran-siv** [comment]
My temporary solution for regular id. Maybe will help to someone
```
async findOneById(id: number): Promise<ProductCategoryEntity | undefined> {
    const category = await this.findOne(id)
    if (!category) {
      return category
    }
    const currentWithChildren = await this.findDescendantsTree(category)
    const currentWithParent = await this.findAncestorsTree(category)
    return {
      ...currentWithChildren,
      parent: currentWithParent.parent,
    }
  }

  async updateMpath(id: number, parentId: number | null | undefined, entityManager?: EntityManager): Promise<void> {
    const category = await this.findOneById(id)
    const parentCategory = parentId && await this.findOneById(parentId)

    const newMPath = parentCategory ? this.getParentList(parentCategory).reverse() : []
    const oldMPath = this.getParentList(category!).reverse()

    const result = [...newMPath, ...oldMPath.slice(oldMPath.indexOf(id))]
    const joined = `${result.join('.')}.`
    const joinedOld = `${oldMPath.join('.')}.`

    const toUpdate = this.getChildList(category!.children)
    toUpdate.push(id)

    const entity = entityManager || this
    const query = `UPDATE product_category SET mpath = REPLACE(mpath, '${joinedOld}', '${joined}')  WHERE product_category.id IN (${toUpdate.join(',')})`
    entity.query(query)
  }

  private getParentList(entity: ProductCategoryEntity): number[] {
    if (!entity.id) {
      return []
    }
    let result: number[] = []
    result.push(entity.id)
    if (entity.parent) {
      result = [...result, ...this.getParentList(entity.parent)]
    }
    return result
  }

  private getChildList(entities: ProductCategoryEntity[]): number[] {
    let result: number[] = []
    for (const entity of entities) {
      result.push(entity.id)
      if (entity.children) {
        result = [...result, ...this.getChildList(entity.children)]
      }
    }
    return result
  }
```

**Comment by gukanozadze** [comment]
Give us the solution guys. At least give us possibility to manually update mpath. this is very frustrating.

**Comment by pvgn** [comment]
Same behaviour on Postgres.

**Comment by Arturokin12** [comment]
Still happening... any solution already?

**Comment by doornot** [comment]
Same behaviour on sqlite3.

**Comment by phil294** [comment]
this is already fixed, check your versions

**Comment by doornot** [comment]
> this is already fixed, check your versions

doesn't working on bulk set, for example:

```ts
const { ids, parentId } = d;
const dashboardRepository = getRepository(DashboardModel);
const parent = await dashboardRepository.findOne({ id: parentId });
const res = await getConnection()
  .createQueryBuilder()
  .update(DashboardModel)
  .set({ parent })
  .where({ id: In(ids) })
  .execute();
return res;
```
